### PR TITLE
[W-9495254, W-9545212] Toggle validation rule via custom settings

### PR DESCRIPTION
**Pull request by vmarasow**
# Changes
We provide the admin with the ability to enable/disable the Application Window Validation trigger via custom settings

# Issues Closed
[W-9495254](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07AH000000PACkYAO/view) - R&A
[W-9545212](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07AH000000Q7DlYAK/view) - EDA

# New Metadata
## Apex
- ApplicationSettingsController
- ApplicationSettingsController_TEST
- ApplicationSettingsMapper
- ApplicationSettingsMapper_TEST
- ApplicationSettingsModel
- ApplicationSettingsModel_TEST
- ApplicationSettingsService
- ApplicationSettingsService_TEST
- ApplicationSettingsVMapper
- ApplicationSettingsVMapper_TEST
- ApplicationSettingsVModel
- ApplicationSettingsVModel_TEST
## LWC
- applicationSettings
## Custom Labels 
- stgApplicationSettingsTitle
- stgApplicationWindowValidation
- stgApplicationWindowValidationDescription
- stgApplicationsNav
## Custom Setting
- Hierarchy_Settings__c.Application_Window_Validation__c

# Deleted Metadata

# Testing Notes

### CODE REVIEW

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
There are a few changes requested here that are required.

- We do not have unit tests for when the new setting is disabled in the trigger.
- There is an agreement issue for the ids in the nav and on the settings page.

The larger highlight for the preferred changes are:
- The API versions of some of the classes are not on 52.0, but 51.0
- There is a use of a static variable on the Application_TDTM Trigger handler which I would prefer to become a configuration load in the handleUpdate and handleInsert methods to use UTIL_CustomSettingsFacade's cacheing, keeping in line with it being an authoritative source for cacheing. I realize there are examples counter to that in EDA's legacy code, but this is the direction we want to move in.

File: force-app/main/default/classes/Application_TDTM.cls [hunk 1]
```diff
@@ -71,6 +77,9 @@ public with sharing class Application_TDTM extends TDTM_Runnable {
      ********************************************************************************************************/
     @TestVisible
     private void handleBeforeInsert(List<SObject> newSObjectList) {
+        if (!isApplicationWindowValidationEnabled) {
```
[the remark this answers is not in the document]
Comment (vmarasow) [reply force-app/main/default/classes/Application_TDTM.cls]: I'm not sure I follow what you are suggesting here. Currently, there is a test covering this.

File: force-app/main/default/classes/ApplicationSettingsController_TEST.cls [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    private static void getRelationshipSettingsVModelException() {
+        ApplicationSettingsVMapper.instance = new STUB_ExceptionApplicationSettingsVMapper();
+
+        try {
+            Test.startTest();
+            ApplicationSettingsVModel applicationSettingsVModel = ApplicationSettingsController.getApplicationSettingsVModel();
+            Test.stopTest();
+        } catch (AuraHandledException e) {
+            System.assertEquals('Script-thrown exception', e.getMessage(), 'Should throw exception.');
+        }
```
Comment (RadSFDeveloper) [inline force-app/main/default/classes/ApplicationSettingsController_TEST.cls:85]: This method will pass w/o exception. Suggesting to store exception in the actual exception var and verify if outside of the try/catch block.
```suggestion

        Exception actualException;
        try {
            Test.startTest();
            ApplicationSettingsVModel applicationSettingsVModel = ApplicationSettingsController.getApplicationSettingsVModel();
            Test.stopTest();
        } catch (AuraHandledException e) {
             actualException = e;
        }
        
        System.assertNotEquals(null, actualException, 'Should throw an exception.');
        System.assertEquals('Script-thrown exception', actualException.getMessage(), 'Exception message should match.');
```

File: force-app/main/default/lwc/applicationSettings/applicationSettings.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    };
+
+    @wire(getApplicationSettingsVModel)
+    wiredApplicationSettings(response) {
+        this._wiredApplicationSettings = response;
+        if (response.data) {
+            this.applicationSettingsVModel = response.data;
+        }
+        if (response.error) {
+            //console.error(response.error);
```
Comment (RadSFDeveloper) [inline force-app/main/default/lwc/applicationSettings/applicationSettings.js:35]: I see other *settings component do not display/log an error either. Wondering if errors should be handled in a future WI for EDA Cloud Settings?

Applications settings when data model retrieval generates an exception which results in no data display. Clicking on [Edit] and following through generates an error in console..

https://user-images.githubusercontent.com/28743286/127585482-9caf3d75-05f5-498a-baf8-985d4a444647.mp4

**Review by RadSFDeveloper — APPROVED** [review]
Nice work on the EDA Settings @vmarasow . There is one comment for a test and suggestion to change, and there is a generic observation on the error handling.

File: force-app/main/default/classes/ApplicationSettingsController_TEST.cls (hunk 2 above)
Comment (RadSFDeveloper) [reply force-app/main/default/classes/ApplicationSettingsController_TEST.cls:85]: The reason why a generic message 'Script-throw exception' is generated is that `AuraHandledException` message is not set. In the case you would like to return exception `e.getMessage()` to the calling component, then need to `setMessage` (see [AuraEnabledUtility](https://github.com/SalesforceFoundation/omega/blob/main/force-app/main/default/classes/support/AuraEnabledUtility.cls#L9) on details). @MatthewBlanski , @xinnanlisf

File: force-app/main/default/classes/Application_TDTM.cls (hunk 1 above)
[the remark this answers is not in the document]
Comment (MatthewBlanski) [reply force-app/main/default/classes/Application_TDTM.cls]: Those lines were previously showing a lack of code coverage when I ran tests at that time. It seems to be resolved now.

File: force-app/main/default/classes/Application_TDTM.cls [hunk 4]
```diff
@@ -71,6 +71,9 @@ public with sharing class Application_TDTM extends TDTM_Runnable {
      ********************************************************************************************************/
     @TestVisible
     private void handleBeforeInsert(List<SObject> newSObjectList) {
+        if (!applicationWindowValidationIsEnabled()) {
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/Application_TDTM.cls:74]:
```suggestion
        if (thisapplicationWindowValidationIsEnabled() == false) {
```

File: force-app/main/default/classes/Application_TDTM.cls [hunk 5]
```diff
@@ -88,6 +91,9 @@ public with sharing class Application_TDTM extends TDTM_Runnable {
      ********************************************************************************************************/
     @TestVisible
     private void handleBeforeUpdate(List<SObject> oldSObjectList, List<SObject> newSObjectList) {
+        if (!applicationWindowValidationIsEnabled()) {
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/Application_TDTM.cls:94]:
```suggestion
        if (this.applicationWindowValidationIsEnabled() == false) {
```

File: force-app/main/default/classes/Application_TEST.cls [hunk 6]
```diff
@@ -45,6 +45,9 @@ private with sharing class Application_TEST {
      *********************************************************/
```
[the remark this answers is not in the document]
Comment (MatthewBlanski) [reply force-app/main/default/classes/Application_TEST.cls:45]: We are still missing the test for Line 58-60 of Application_TDTM.

File: force-app/main/default/lwc/applicationSettings/applicationSettings.js (hunk 3 above)
Comment (MatthewBlanski) [reply force-app/main/default/lwc/applicationSettings/applicationSettings.js:35]: We are handling errors in this major release and the next one. The handling got descoped from the prior one.

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
Some very minor changes on a single unit test that is missing for 100% meaningful code coverage on Application_TDTM and a styling nitpick on using this. for some method calls. The latter of these can simply be incorporated per the suggestions.

File: force-app/main/default/classes/ApplicationSettingsController_TEST.cls (hunk 2 above)
Comment (MatthewBlanski) [reply force-app/main/default/classes/ApplicationSettingsController_TEST.cls:85]: I'm placing an alternative here for posterity. Having the assert cause a test failure will also meet the need.
```suggestion

        try {
            Test.startTest();
            ApplicationSettingsVModel applicationSettingsVModel = ApplicationSettingsController.getApplicationSettingsVModel();
            Test.stopTest();
            System.assert(false,'An AuraHandledException should be thrown.');
        } catch (AuraHandledException e) {
            System.assertEquals('Script-thrown exception', e.getMessage(), 'Should throw exception.');
        }
```

**Review by MatthewBlanski — APPROVED** [review]

**Review by RadSFDeveloper — APPROVED** [review]
Looks good to me.

**Review by xinnanlisf — APPROVED** [review]
Nicely done! Approved!

**Comment by ravensisco** [comment]
Testing has been completed from the Admissions Connect side.

**Comment by rhakeshj** [comment]
QA passed

**Pull request merged**